src/relay/backend/build_module.cc

```diff
@@ -244,6 +244,9 @@ class RelayBuildModule : public runtime::ModuleNode {
       GlobalVar main_glb_var = relay_module->GetGlobalVar("main");
       Function main_func = Downcast<Function>(relay_module->Lookup(main_glb_var));
       auto new_main = BindParamsByName(main_func, params);
+      // copy module to avoid changing our input
+      relay_module = IRModule(relay_module->functions, relay_module->type_definitions,
```

Comment:
If I understand the issue correctly, I think we need to call CopyOnWrite instead.

IRModule is actually defined with `CopyOnWrite` method [here](https://github.com/apache/incubator-tvm/blob/master/include/tvm/ir/module.h#L332), the macro expands to [this](https://github.com/apache/incubator-tvm/blob/master/include/tvm/runtime/object.h#L755-L763). What we need to do is just call its `CopyOnWrite` method to generate a unique copy of IRModule.

Reply:
OK, so I changed it to call `CopyOnWrite` and it appears to work.

To understand this better, I got the re-instatiation from `FunctionPassNode::operator()` here:
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/ir/transform.cc#L123
is there a reason that isn't done via `CopyOnWrite`?
Superficially, it seems that the update is then 
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/ir/transform.cc#L135
which would seem very similar to the `Update` done in the `Optimize`
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/relay/backend/build_module.cc#L247
which itself just calls add
https://github.com/apache/incubator-tvm/blob/99745a44407f2d1bd06b8c6a47e6c6c5239ec665/src/ir/module.cc#L271-L273

Is  something that could also use `CopyOnWrite` or is it not applicable there.

Reply:
I must admit I'm not understanding this very well yet, in particular in for the use of the new `relay_module_ptr` I get from `CopyOnWrite` and the original module.
Which one would I use for what after CopyOnWrite?